x-pack/plugins/security/server/lib/authorization/has_privileges.test.js

```diff
@@ -0,0 +1,311 @@
+  const privileges = [privilege];
+  const result = await hasPrivileges(privileges);
+
+  expectNoDeprecationLogged(mockServer);
+  expect(mockCallWithRequest).toHaveBeenCalledWith(request, 'shield.hasPrivileges', {
+    body: {
+      applications: [{
+        application: defaultApplication,
+        resources: [DEFAULT_RESOURCE],
+        privileges: [
+          getVersionPrivilege(defaultVersion), getLoginPrivilege(), ...privileges
+        ]
+      }]
+    }
+  });
+  expect(result).toEqual({
+    result: HAS_PRIVILEGES_RESULT.AUTHORIZED,
+    username,
+    missing: [],
+  });
+});
+
+test(`returns unauthorized they have only one application privilege`, async () => {
+  const privilege1 = `action:saved_objects/${savedObjectTypes[0]}/get`;
+  const privilege2 = `action:saved_objects/${savedObjectTypes[0]}/create`;
+  const username = 'foo-username';
+  const mockServer = createMockServer();
+  const mockCallWithRequest = createMockCallWithRequest([
+    mockApplicationPrivilegeResponse({
+      hasAllRequested: false,
+      privileges: {
+        [getVersionPrivilege(defaultVersion)]: true,
+        [getLoginPrivilege()]: true,
+        [privilege1]: true,
+        [privilege2]: false,
+      },
+      application: defaultApplication,
+      username,
+    })
+  ]);
+
+  const hasPrivilegesWithRequest = hasPrivilegesWithServer(mockServer);
+  const request = Symbol();
+  const hasPrivileges = hasPrivilegesWithRequest(request);
+  const privileges = [privilege1, privilege2];
+  const result = await hasPrivileges(privileges);
+
+  expectNoDeprecationLogged(mockServer);
+  expect(mockCallWithRequest).toHaveBeenCalledWith(request, 'shield.hasPrivileges', {
+    body: {
+      applications: [{
+        application: defaultApplication,
+        resources: [DEFAULT_RESOURCE],
+        privileges: [
+          getVersionPrivilege(defaultVersion), getLoginPrivilege(), ...privileges
+        ]
+      }]
+    }
+  });
+  expect(result).toEqual({
+    result: HAS_PRIVILEGES_RESULT.UNAUTHORIZED,
+    username,
+    missing: [privilege2],
+  });
+});
+
+test(`throws error if missing version privilege and has login privilege`, async () => {
+  const privilege = `action:saved_objects/${savedObjectTypes[0]}/get`;
+  const mockServer = createMockServer();
+  createMockCallWithRequest([
+    mockApplicationPrivilegeResponse({
+      hasAllRequested: false,
+      privileges: {
+        [getVersionPrivilege(defaultVersion)]: false,
+        [getLoginPrivilege()]: true,
+        [privilege]: true,
+      }
+    })
+  ]);
+
+  const hasPrivilegesWithRequest = hasPrivilegesWithServer(mockServer);
+  const hasPrivileges = hasPrivilegesWithRequest({});
+
+  await expect(hasPrivileges([privilege])).rejects.toThrowErrorMatchingSnapshot();
+  expectNoDeprecationLogged(mockServer);
+});
+
+describe('legacy fallback with no application privileges', () => {
+  test(`returns unauthorized if they have no privileges on the kibana index`, async () => {
```

Comment:
I think we need a test for `returns unauthorized if they only have irrelevant privileges on the kibana index`. Currently your tests are only verifying the behavior for "we have relevant privileges" and "we don't have relevant privileges", but you don't verify the behavior of when a user has some legacy privileges to the index that just don't happen to be the ones that Kibana cares about. We can't test every possible combination of privileges, but we should at least test those buckets.

Reply:
When we call the `_has_privileges` endpoint in Elasticsearch, we specify the privileges that we wish to have them check, so the "irrelevant" ones will never be in the response because we never asked for them to be considered.